README.md
# Import Once [![npm version](https://badge.fury.io/js/node-sass-import-once.svg)](http://badge.fury.io/js/node-sass-import-once) [![Build Status](https://travis-ci.org/at-import/node-sass-import-once.svg)](https://travis-ci.org/at-import/node-sass-import-once)

General `import-once` importer for [`node-sass`](https://github.com/sass/node-sass) inspired by [Eyeglass](https://github.com/sass-eyeglass/eyeglass).

## Installation

```bash
$ npm install node-sass-import-once --save-dev
```

**Requires [Node Sass](https://github.com/sass/node-sass) >= v3.0.0-alpha.5**

## Usage

```javascript
var sass = require('node-sass');,
    importOnce = require('node-sass-import-once');

sass.render({
  file: scss_filename,
  importer: importOnce,
  importOnce: {
    index: false,
    css: false,
    bower: false
  }
});
```

If you are using Gulp or Grunt to compile (or similiar), require `node-sass-import-once` as normal and pass it as the `importer` option in to your task.

## Features

Node Sass Import Once will ensure that each file import is only imported once. In addition to that, the following options are available:

* **index** - Allows folders to contain an index file (`index.scss`, `index.sass`, `_index.scss`, `_index.sass`) that will automatically be imported if just the folder name is imported. For instance, `@import 'partials';` will try and import `(_)partials.s(c|a)ss` first, then `partials/(_)index.s(c|a)ss`.
* **css** - Allows a CSS file to be imported directly into your Sass files. Simply don't include a file extension, and if it's available as a CSS file it'll be imported!
* **bower** - Automatically parses your [Bower](http://bower.io/) directory for the files you're requested. Will look for the file in `bower_components` (or specified directory from `.bowerrc`), `bower_components/{module-name}/stylesheets`, `bower_components/{module-name}-sass/stylesheets`, `bower_components/sass-{module-name}/stylesheets`, `bower_components/{module-name}/sass`, `bower_components/{module-name}-sass/sass`, `bower_components/sass-{module-name}/sass`. `{module-name}` is the first section of your import path, so in `@import 'breakpoint'` **breakpoint** would be your module name and in `@import 'toolkit/kickstart` **toolkit** would be your module name. If you also have **css** enabled, it will look inside of `bower_components/{module-name}/dist` and `bower_components/{module-name}/dist/css`.


CHANGELOG.md
# Import Once Changelog

## v1.1.0
### March 26, 2015
* **New** Added support for files passed in as strings instead of as file paths
* **New** Added Changelog

## v1.0.0
### April 6, 2015
* **New** Initial Release


tests/main.js
'use strict';

var should = require('should'),
    sass = require('node-sass'),
    path = require('path'),
    fs = require('fs'),
    importer = require('../index');

var filePath = function (file) {
  return path.join(path.join(__dirname, 'sass'), file);
};

describe('import-once', function () {
  it('should import files only once', function (done) {
    var file = filePath('basic-import-once.scss'),
        expectedIncludes = [
          file,
          'partial-with-selectors'
        ];

    sass.render({
      'file': file,
      'importer': importer
    }, function (err, result) {
      if (err) {
        throw err;
      }
      should.exist(result);
      result.stats.includedFiles.should.eql(expectedIncludes);
      String(result.css).should.equal(
        fs.readFileSync(path.join(__dirname, 'css/basic-import-once.css'), 'utf8')
      );
      done();
    });
  });

  it('should import `index` files from a folder', function (done) {
    var file = filePath('import-index.scss'),
        expectedIncludes = [
          file,
          'foo'
        ];

    sass.render({
      'file': file,
      'importer': importer,
      'importOnce': {
        'index': true
      }
    }, function (err, result) {
      if (err) {
        throw err;
      }
      should.exist(result);
      result.stats.includedFiles.should.eql(expectedIncludes);
      String(result.css).should.equal(
        fs.readFileSync(path.join(__dirname, 'css/imported-index.css'), 'utf8')
      );
      done();
    });
  });

  it('should import `css` files as Sass from a folder', function (done) {
    var file = filePath('import-css.scss'),
        expectedIncludes = [
          file,
          'imported-css'
        ];

    sass.render({
      'file': file,
      'importer': importer,
      'importOnce': {
        'css': true
      }
    }, function (err, result) {
      if (err) {
        throw err;
      }
      should.exist(result);
      result.stats.includedFiles.should.eql(expectedIncludes);
      // console.log(result.css.toString());
      String(result.css).should.equal(
        fs.readFileSync(path.join(__dirname, 'css/imported-css.css'), 'utf8')
      );
      done();
    });
  });

  it('should import `bower` files as Sass from a folder', function (done) {
    var file = filePath('import-bower.scss'),
        expectedIncludes = [
          file,
          'breakpoint',
          'breakpoint/context',
          'breakpoint/helpers',
          'breakpoint/no-query',
          'breakpoint/parsers',
          'breakpoint/respond-to',
          'double/default',
          'double/default-pair',
          'double/double-string',
          'parsers/double',
          'parsers/query',
          'parsers/resolution',
          'parsers/single',
          'parsers/triple',
          'resolution/resolution',
          'single/default',
          'toolkit/kickstart',
          'triple/default'
        ];

    sass.render({
      'file': file,
      'importer': importer,
      'importOnce': {
        'bower': true
      }
    }, function (err, result) {
      if (err) {
        throw err;
      }
      should.exist(result);
      result.stats.includedFiles.should.eql(expectedIncludes);
      // console.log(result.css.toString());
      String(result.css).should.equal(
        fs.readFileSync(path.join(__dirname, 'css/imported-bower.css'), 'utf8')
      );
      done();
    });
  });

  it('should support custom include paths', function (done) {
    var file = filePath('import-custom.scss'),
        expectedIncludes = [
          file,
          'custom-import'
        ];

    sass.render({
      'file': file,
      'importer': importer,
      'includePaths': [
        'custom'
      ]
    }, function (err, result) {
      if (err) {
        throw err;
      }
      should.exist(result);
      // console.log(result.stats.includedFiles);
      result.stats.includedFiles.should.eql(expectedIncludes);
      // console.log(result.css.toString());
      String(result.css).should.equal(
        fs.readFileSync(path.join(__dirname, 'css/imported-custom.css'), 'utf8')
      );
      done();
    });
  });

  it('should import Bootstrap as Sass', function (done) {
    var file = filePath('import-bootstrap.scss'),
        expectedIncludes = [
          file,
          'bootstrap'
        ];

    sass.render({
      'file': file,
      'importer': importer,
      'importOnce': {
        'css': true,
        'bower': true
      }
    }, function (err, result) {
      if (err) {
        throw err;
      }
      should.exist(result);
      result.stats.includedFiles.should.eql(expectedIncludes);
      // console.log(result.css.toString());
      String(result.css).should.equal(
        fs.readFileSync(path.join(__dirname, 'css/imported-bootstrap.css'), 'utf8')
      );
      done();
    });
  });

  it('should fall back to import paths and bower if data is passed in instead of a file name', function (done) {
    var file = filePath('basic-import-once.scss'),
        expectedIncludes = [
          'partial-with-selectors'
        ];

    sass.render({
      'data': fs.readFileSync(file, 'utf-8'),
      'importer': importer,
      'includePaths': [
        path.dirname(file)
      ]
    }, function (err, result) {
      if (err) {
        throw err;
      }
      should.exist(result);
      result.stats.includedFiles.should.eql(expectedIncludes);
      String(result.css).should.equal(
        fs.readFileSync(path.join(__dirname, 'css/basic-import-once.css'), 'utf8')
      );
      done();
    });
  });
});


import-once.js
/**
  * Shameless adapted from [Eyeglass](https://github.com/sass-eyeglass/eyeglass)
  *  because I wanted a general-use import-once importer for Node
**/
'use strict';

var fs = require('fs'),
    path = require('path');

/**
 * All imports use the forward slash as a directory
 * delimeter. This function converts to the filesystem's
 * delimeter if it uses an alternate.
**/
function makeFsPath(importPath) {
  var fsPath = importPath;
  if (path.sep !== '/') {
    fsPath = fsPath.replace(/\//, path.sep);
  }
  return fsPath;
}

/**
  * Determines if a file should be imported or not
**/
function importOnce(data, done) {
  if (this._importOnceCache[data.file]) {
    done({
      contents: '',
      filename: 'already-imported:' + data.file
    });
  } else {
    this._importOnceCache[data.file] = true;
    done(data);
  }
}

/**
 * Sass imports are usually in an abstract form in that
 * they leave off the partial prefix and the suffix.
 * This code creates the possible extensions, whether it is a partial
 * and whether it is a directory index file having those
 * same possible variations. If the import contains an extension,
 * then it is left alone.
 *
**/
function getFileNames(abstractName) {
  var names = [];
  if (path.extname(abstractName)) {
    names.push(abstractName);
  } else {
    var directory = path.dirname(abstractName);
    var basename = path.basename(abstractName);

    // Standard File Names
    ['', '_'].forEach(function(prefix) {
      ['.scss', '.sass'].forEach(function(ext) {
        names.push(path.join(directory, prefix + basename + ext));
      });
    });

    // Index Files
    if (this.options.importOnce.index) {
      ['', '_'].forEach(function(prefix) {
        ['.scss', '.sass'].forEach(function(ext) {
          names.push(path.join(abstractName, prefix + 'index' + ext));
        });
      });
    }

    // CSS Files
    if (this.options.importOnce.css) {
      names.push(abstractName + '.css');
    }

  }
  return names;
}

/**
  * Build list of potential Bower imports
**/
function getBowerNames(uri) {
  var gfn = getFileNames.bind(this);

  var bowerrc = path.resolve(process.cwd(), '.bowerrc'),
      bowerPath = 'bower_components',
      core = uri.split('/')[0],
      paths = [],
      results = [];

  uri = makeFsPath(uri);

  if (fs.existsSync(bowerrc)) {
    bowerrc = JSON.parse(fs.readFileSync(bowerrc, 'utf-8'));
    if (bowerrc.directory) {
      bowerPath = bowerrc.directory;
    }
  }

  // Resolve the path to the Bower repository;
  bowerPath = path.resolve(process.cwd(), bowerPath);

  // Basic, add import path-esque
  paths.push(path.resolve(bowerPath, uri));
  // For those projects that were Ruby gems and are now distributed through Bower
  if (core !== '..' && core !== '.') {
    paths.push(path.resolve(bowerPath, core, 'stylesheets', uri));
    paths.push(path.resolve(bowerPath, core + '-sass', 'stylesheets', uri));
    paths.push(path.resolve(bowerPath, 'sass-' + core, 'stylesheets', uri));

    paths.push(path.resolve(bowerPath, core, 'sass', uri));
    paths.push(path.resolve(bowerPath, core + '-sass', 'sass', uri));
    paths.push(path.resolve(bowerPath, 'sass-' + core, 'sass', uri));

    if (this.options.importOnce.css) {
      paths.push(path.resolve(bowerPath, core, 'dist', uri));
      paths.push(path.resolve(bowerPath, core, 'dist', 'css', uri));
    }
  }

  // Get the file names for all of the paths!
  paths.forEach(function(path) {
    results = results.concat(gfn(path));
  });

  return results;
}

/**
  * getIn
**/
function getIncludePaths(uri) {
  // From https://github.com/sass/node-sass/issues/762#issuecomment-80580955
  var arr = this.options.includePaths.split(path.delimiter),
      gfn = getFileNames.bind(this),
      paths = [];

  arr.forEach(function(includePath) {
    paths = paths.concat(gfn(path.resolve(process.cwd(), includePath, uri)));
  });

  return paths;
}

// This is a bootstrap function for calling readFirstFile.
function readAbstractFile(uri, abstractName, cb) {
  var gfn = getFileNames.bind(this),
      gip = getIncludePaths.bind(this),
      gbn = getBowerNames.bind(this),
      css = this.options.importOnce.css;

  var files = gfn(abstractName);

  if (this.options.includePaths) {
    files = files.concat(gip(uri));
  }

  if (this.options.importOnce.bower) {
    files = files.concat(gbn(uri));
  }

  readFirstFile(uri, files, css, cb);
}

/**
 * Asynchronously walks the file list until a match is found. If
 * no matches are found, calls the callback with an error
**/
function readFirstFile(uri, filenames, css, cb, examinedFiles) {
  var filename = filenames.shift();
  examinedFiles = examinedFiles || [];
  examinedFiles.push(filename);
  fs.readFile(filename, function(err, data) {
    if (err) {
      if (filenames.length) {
        readFirstFile(uri, filenames, css, cb, examinedFiles);
      } else {
        cb(new Error('Could not import `' + uri + '` from any of the following locations:\n  ' + examinedFiles.join('\n  ')));
      }
    } else {
      cb(null, {
        contents: data.toString(),
        file: filename
      });
    }
  });
}

/**
  * Import the goodies!
**/
function importer(uri, prev, done) {
  var isRealFile = fs.existsSync(prev),
      io = importOnce.bind(this),
      raf = readAbstractFile.bind(this),
      file;

  // Ensure options are available
  if (!this.options.importOnce) {
    this.options.importOnce = {};
  }

  // Set default index import
  if (!this.options.importOnce.index) {
    this.options.importOnce.index = false;
  }

  // Set default bower import
  if (!this.options.importOnce.bower) {
    this.options.importOnce.bower = false;
  }

  // Set default css import
  if (!this.options.importOnce.css) {
    this.options.importOnce.css = false;
  }

  // Create an import cache if it doesn't exist
  if (!this._importOnceCache) {
    this._importOnceCache = {};
  }

  if (isRealFile) {
    file = path.resolve(path.dirname(prev), makeFsPath(uri));
    raf(uri, file, function (err, data) {
      if (err) {
        console.log(err.toString());
        done({});
      } else {
        io(data, done);
      }
    });
  }
}

/**
  * Exports the importer
**/
module.exports = importer;


index.js
/**
  * Shameless adapted from [Eyeglass](https://github.com/sass-eyeglass/eyeglass)
  *  because I wanted a general-use import-once importer for Node
**/
'use strict';

var fs = require('fs'),
    path = require('path');

/**
 * All imports use the forward slash as a directory
 * delimeter. This function converts to the filesystem's
 * delimeter if it uses an alternate.
**/
var makeFsPath = function makeFsPath(importPath) {
  var fsPath = importPath;
  if (path.sep !== '/') {
    fsPath = fsPath.replace(/\//, path.sep);
  }
  return fsPath;
};

/**
  * Determines if a file should be imported or not
**/
var importOnce = function importOnce(data, done) {
  if (this._importOnceCache[data.file]) {
    done({
      'contents': '',
      'filename': 'already-imported:' + data.file
    });
  }
  else {
    this._importOnceCache[data.file] = true;
    done(data);
  }
};

/**
 * Sass imports are usually in an abstract form in that
 * they leave off the partial prefix and the suffix.
 * This code creates the possible extensions, whether it is a partial
 * and whether it is a directory index file having those
 * same possible variations. If the import contains an extension,
 * then it is left alone.
 *
**/
var getFileNames = function getFileNames(abstractName) {
  var names = [],
      directory,
      basename;

  if (path.extname(abstractName)) {
    names.push(abstractName);
  }
  else {
    directory = path.dirname(abstractName);
    basename = path.basename(abstractName);

    // Standard File Names
    [ '', '_' ].forEach(function(prefix) {
      [ '.scss', '.sass' ].forEach(function(ext) {
        names.push(path.join(directory, prefix + basename + ext));
      });
    });

    // Index Files
    if (this.options.importOnce.index) {
      [ '', '_' ].forEach(function(prefix) {
        [ '.scss', '.sass' ].forEach(function(ext) {
          names.push(path.join(abstractName, prefix + 'index' + ext));
        });
      });
    }

    // CSS Files
    if (this.options.importOnce.css) {
      names.push(abstractName + '.css');
    }
  }
  return names;
};

/**
  * Build list of potential Bower imports
**/
var getBowerNames = function getBowerNames(uri) {
  var gfn = getFileNames.bind(this);

  var bowerrc = path.resolve(process.cwd(), '.bowerrc'),
      bowerPath = 'bower_components',
      core = uri.split('/')[0],
      paths = [],
      results = [];

  uri = makeFsPath(uri);

  if (fs.existsSync(bowerrc)) {
    bowerrc = JSON.parse(fs.readFileSync(bowerrc, 'utf-8'));
    if (bowerrc.directory) {
      bowerPath = bowerrc.directory;
    }
  }

  // Resolve the path to the Bower repository;
  bowerPath = path.resolve(process.cwd(), bowerPath);

  // Basic, add import path-esque
  paths.push(path.resolve(bowerPath, uri));
  // For those projects that were Ruby gems and are now distributed through Bower
  if (core !== '..' && core !== '.') {
    paths.push(path.resolve(bowerPath, core, uri));
    paths.push(path.resolve(bowerPath, core, 'stylesheets', uri));
    paths.push(path.resolve(bowerPath, core + '-sass', 'stylesheets', uri));
    paths.push(path.resolve(bowerPath, 'sass-' + core, 'stylesheets', uri));

    paths.push(path.resolve(bowerPath, core, 'sass', uri));
    paths.push(path.resolve(bowerPath, core + '-sass', 'sass', uri));
    paths.push(path.resolve(bowerPath, 'sass-' + core, 'sass', uri));

    if (this.options.importOnce.css) {
      paths.push(path.resolve(bowerPath, core, 'dist', uri));
      paths.push(path.resolve(bowerPath, core, 'dist', 'css', uri));
    }
  }

  // Get the file names for all of the paths!
  paths.forEach(function(pathName) {
    results = results.concat(gfn(pathName));
  });

  return results;
};

/**
  * getIn
**/
var getIncludePaths = function getIncludePaths(uri) {
  // From https://github.com/sass/node-sass/issues/762#issuecomment-80580955
  var arr = this.options.includePaths.split(path.delimiter),
      gfn = getFileNames.bind(this),
      paths = [];

  arr.forEach(function(includePath) {
    paths = paths.concat(gfn(path.resolve(process.cwd(), includePath, uri)));
  });

  return paths;
};

/**
 * Asynchronously walks the file list until a match is found. If
 * no matches are found, calls the callback with an error
**/
var readFirstFile = function readFirstFile(uri, filenames, css, cb, examinedFiles) {
  var filename = filenames.shift();
  examinedFiles = examinedFiles || [];
  examinedFiles.push(filename);
  fs.readFile(filename, function(err, data) {
    if (err) {
      if (filenames.length) {
        readFirstFile(uri, filenames, css, cb, examinedFiles);
      }
      else {
        cb(new Error('Could not import `' + uri + '` from any of the following locations:\n  ' + examinedFiles.join('\n  ')));
      }
    }
    else {
      cb(null, {
        'contents': data.toString(),
        'file': filename
      });
    }
  });
};

// This is a bootstrap function for calling readFirstFile.
var readAbstractFile = function readAbstractFile(uri, abstractName, cb) {
  var gfn = getFileNames.bind(this),
      gip = getIncludePaths.bind(this),
      gbn = getBowerNames.bind(this),
      css = this.options.importOnce.css;

  var files = gfn(abstractName);

  if (this.options.includePaths) {
    files = files.concat(gip(uri));
  }

  if (this.options.importOnce.bower) {
    files = files.concat(gbn(uri));
  }

  readFirstFile(uri, files, css, cb);
};

/**
  * Import the goodies!
**/
var importer = function importer(uri, prev, done) {
  var isRealFile = fs.existsSync(prev),
      io = importOnce.bind(this),
      raf = readAbstractFile.bind(this),
      file;

  // Ensure options are available
  if (!this.options.importOnce) {
    this.options.importOnce = {};
  }

  // Set default index import
  if (!this.options.importOnce.index) {
    this.options.importOnce.index = false;
  }

  // Set default bower import
  if (!this.options.importOnce.bower) {
    this.options.importOnce.bower = false;
  }

  // Set default css import
  if (!this.options.importOnce.css) {
    this.options.importOnce.css = false;
  }

  // Create an import cache if it doesn't exist
  if (!this._importOnceCache) {
    this._importOnceCache = {};
  }

  if (isRealFile) {
    file = path.resolve(path.dirname(prev), makeFsPath(uri));
    raf(uri, file, function (err, data) {
      if (err) {
        console.log(err.toString());
        done({});
      }
      else {
        io(data, done);
      }
    });
  }
  else {
    raf(uri, process.cwd(), function (err, data) {
      if (err) {
        console.log(err.toString());
        done({});
      }
      else {
        io(data, done);
      }
    });
  }
};

/**
  * Exports the importer
**/
module.exports = importer;
